manim/grpc/impl/frame_server_impl.py

```diff
@@ -0,0 +1,262 @@
+from ...config import camera_config
+from ...config import file_writer_config
+from ...scene import scene
+from ..gen import frameserver_pb2
+from ..gen import frameserver_pb2_grpc
+from ..gen import renderserver_pb2
+from ..gen import renderserver_pb2_grpc
+from concurrent import futures
+from google.protobuf import json_format
+from watchdog.events import LoggingEventHandler, FileSystemEventHandler
+from watchdog.observers import Observer
+import grpc
+import subprocess as sp
+import threading
+import time
+import ctypes
+from ...utils.module_ops import (
+    get_module,
+    get_scene_classes_from_module,
+    get_scenes_to_render,
+)
+from ...config.logger import logger
+from ...constants import JS_RENDERER_INFO
+
+
+class FrameServer(frameserver_pb2_grpc.FrameServerServicer):
+    def animation_index_is_cached(self, animation_index):
+        return animation_index < len(self.keyframes)
+
+    def __init__(self, server, scene_class):
+        self.server = server
+        self.keyframes = []
+        self.scene = scene_class(self)
+        self.scene_thread = threading.Thread(
+            target=lambda s: s.render(), args=(self.scene,)
+        )
+        self.previous_frame_animation_index = None
+        self.scene_finished = False
+
+        path = "./example_scenes/basic.py"
+        event_handler = UpdateFrontendHandler(self)
+        observer = Observer()
+        observer.schedule(event_handler, path)
+        observer.start()
+
+        # If a javascript renderer is running, notify it of the scene being served. If
+        # not, spawn one and it will request the scene when it starts.
+        with grpc.insecure_channel("localhost:50052") as channel:
+            stub = renderserver_pb2_grpc.RenderServerStub(channel)
+            request = renderserver_pb2.NewSceneRequest(name=str(self.scene))
+            try:
+                stub.NewScene(request)
+            except grpc._channel._InactiveRpcError:
+                logger.warning(f"No frontend was detected at localhost:50052.")
+                try:
+                    sp.Popen(camera_config["js_renderer_path"])
+                except PermissionError:
+                    logger.info(JS_RENDERER_INFO)
+                    self.server.stop(None)
+                    return
+
+        self.scene_thread.start()
+
+    def signal_pending_animation(self, animation_index):
+        self.scene.start_animation = animation_index
+        self.scene.animation_finished.set()
+        return frameserver_pb2.FrameResponse(frame_pending=True)
+
+    def GetFrameAtTime(self, request, context):
```

Comment:
Odd name - why camel case?

Reply:
gRPC requires the name to match the one described in the .proto file [here](https://github.com/ManimCommunity/manim/pull/465/files#diff-93c2a61976cfcd4cc114e97839f15aabR8). gRPC is language-agnostic so from its perspective it's only incidental that the RPC is implemented in python.

Reply:
Thanks for the explanation. Perhaps a comment to that effect wouldn't hurt.